File: py/day17.py
import heapq
import math
from dataclasses import dataclass
from collections import deque, defaultdict, Counter
from typing import *
import time

import numpy as np
import itertools
import re

import fire


def main():
    print('hi!')

    xlim = (257,  286)
    ylim = (-101, -57)

    def hit(pos):
        return (xlim[0] <= pos[0] <= xlim[1]) and (ylim[0] <= pos[1] <= ylim[1])

    def step(pos, vel):
        pos[0] += vel[0]
        pos[1] += vel[1]

        if vel[0] > 0:
            vel[0] -= 1
        elif vel[0] < 0:
            vel[0] += 1
        else:
            vel[0] = 0
        vel[1] -= 1

    def trial(vel):
        pos = [0, 0]
        max_y = pos[1]
        while pos[1] >= ylim[0] and pos[0] <= xlim[1]:
            if pos[1] > max_y:
                max_y = pos[1]
            if hit(pos):
                return pos, max_y
            step(pos, vel)

        return None, max_y

    t0 = time.time()
    max_max_y = 0
    for u in range(1, xlim[1]):
        for v in range(-ylim[0]):
            p, max_y = trial([u, v])
            if p:
                if max_y > max_max_y:
                    max_max_y = max_y

    print('1:', max_max_y, time.time() - t0, 's')

    t1 = time.time()
    hits = 0
    for u in range(1, xlim[1]+1):
        for v in range(ylim[0], -ylim[0]):
            p, max_y = trial([u, v])
            if p:
                hits += 1

    print('2:', hits, time.time() - t1, 's')


if __name__ == '__main__':
    fire.Fire(main)

File: go/day17.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"regexp"
	"runtime/debug"
	"strconv"
	"time"

	"./util"
)

func main() {
	defer func() {
		if r := recover(); r != nil {
			fmt.Println(string(debug.Stack()))
		}
	}()

	var filename string = os.Args[1]
	t0 := time.Now()
	fmt.Printf("Reading from %s,", filename)
	file, err := os.Open(filename)
	if err != nil {
		log.Panicln(err)
	}
	defer file.Close()

	re := regexp.MustCompile(`^target area: x=(-?[0-9].*)\.\.(-?[0-9].*), y=(-?[0-9].*)\.\.(-?[0-9].*)$`)
	scanner := bufio.NewScanner(file)
	scanner.Scan()
	groups := re.FindStringSubmatch(scanner.Text())
	var ints []int
	for _, m := range groups[1:] {
		v, err := strconv.Atoi(m)
		if err != nil {
			log.Fatalln(err)
		}
		ints = append(ints, v)
	}
	x0, x1, y0, y1 := ints[0], ints[1], ints[2], ints[3]

	if err := scanner.Err(); err != nil {
		log.Panicln(err)
	}
	fmt.Printf(" took %v\n", time.Since(t0))
	fmt.Println(x0, x1, y0, y1)

	step := func(pos *util.Point2D, vel *util.Point2D) {
		pos.X += vel.X
		pos.Y += vel.Y

		if vel.X > 0 {
			vel.X--
		} else if vel.X < 0 {
			vel.X++
		}
		vel.Y--
	}

	trial := func(vel util.Point2D) (bool, int) {
		pos := util.Point2D{X: 0, Y: 0}
		maxY := 0
		for pos.Y >= y0 && pos.X <= x1 {
			if pos.Y > maxY {
				maxY = pos.Y
			}
			if pos.X >= x0 && pos.X <= x1 && pos.Y >= y0 && pos.Y <= y1 {
				return true, maxY
			}
			step(&pos, &vel)
		}
		return false, maxY
	}

	t1 := time.Now()
	t := 0
	for u := 1; u < x1+1; u++ {
		for v := 0; v < -y0; v++ {
			if hit, maxY := trial(util.Point2D{X: u, Y: v}); hit {
				if maxY > t {
					t = maxY
				}
			}
		}

	}
	fmt.Printf("1: %v, %v\n", t, time.Since(t1))

	t2 := time.Now()
	t = 0
	for u := 1; u < x1+1; u++ {
		for v := y0; v < -y0; v++ {
			if hit, _ := trial(util.Point2D{X: u, Y: v}); hit {
				t++
			}
		}

	}
	fmt.Printf("2: %v, %v\n", t, time.Since(t2))
}
